### PR TITLE
Add App Associated Role Resolver implementation and handle roles claim

### PR DESCRIPTION
**Pull request by SujanSanjula96**
### Proposed changes in this pull request

Add Role Resolver implementation to get the app associated roles of the authenticated user and handle the roles claim in Role V2 runtime.

Related Issue - https://github.com/wso2/product-is/issues/16363

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 1]
```diff
@@ -1098,6 +1110,20 @@ private boolean enableMergingCustomClaimMappingsWithDefaultMappings() {
...
+     * Handle role app association in roles claim.
+     *
+     * @param appAssociatedRoles App associated roles.
+     * @param mappedAttrs Mapped claim attributes.
+     */
+    private void handleRoleAppAssoication(Map<String, String> mappedAttrs, String appAssociatedRoles) {
+
+        if (mappedAttrs.containsKey(getLocalGroupsClaimURI())) {
+            mappedAttrs.put(getLocalGroupsClaimURI(),
+                    StringUtils.isEmpty(appAssociatedRoles) ? "" : appAssociatedRoles);
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:1123]:
```suggestion
                    StringUtils.isEmpty(appAssociatedRoles) ? StringUtils.Empty : appAssociatedRoles);
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java [hunk 2]
```diff
@@ -1048,4 +1053,22 @@ protected void unregisterConfigurationManager(ConfigurationManager configuration
...
+    protected void setRoleManagementServiceV2(RoleManagementService roleManagementService) {
+
+        FrameworkServiceDataHolder.getInstance().setRoleManagementServiceV2(roleManagementService);
+        log.debug("RoleManagementServiceV2 set in ApplicationManagementServiceComponent bundle.");
+    }
+
+    protected void unsetRoleManagementServiceV2(RoleManagementService roleManagementService) {
+
+        FrameworkServiceDataHolder.getInstance().setRoleManagementServiceV2(null);
+        log.debug("RoleManagementServiceV2 unset in ApplicationManagementServiceComponent bundle.");
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java:1072]:
```suggestion
        log.debug("RoleManagementServiceV2 unset in FrameworkServiceComponent bundle.");
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java [hunk 3]
```diff
@@ -1048,4 +1053,22 @@ protected void unregisterConfigurationManager(ConfigurationManager configuration
...
+    @Reference(
+            name = "org.wso2.carbon.identity.role.v2.mgt.core.RoleManagementService",
+            service = org.wso2.carbon.identity.role.v2.mgt.core.RoleManagementService.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetRoleManagementServiceV2")
+    protected void setRoleManagementServiceV2(RoleManagementService roleManagementService) {
+
+        FrameworkServiceDataHolder.getInstance().setRoleManagementServiceV2(roleManagementService);
+        log.debug("RoleManagementServiceV2 set in ApplicationManagementServiceComponent bundle.");
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java:1066]:
```suggestion
        log.debug("RoleManagementServiceV2 set in FrameworkServiceComponent bundle.");
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ */
+public class RoleResolverUtils {
+
+    /**
+     * Throw an ApplicationRolesServerException with the given error message.
+     *
+     * @param error The error enum.
+     * @param e     The throwable.
+     * @param data  The error message data.
+     * @return ApplicationRolesServerException
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java:37]:
```suggestion
     * @return ApplicationRolesServerException.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        }
+        return new ApplicationRolesServerException(error.getMessage(), description, error.getCode(), e);
+    }
+
+    /**
+     * Throw an ApplicationRolesClientException with the given error message.
+     *
+     * @param error The error enum.
+     * @param data The error message data.
+     * @return ApplicationRolesClientException
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java:54]:
```suggestion
     * @return ApplicationRolesClientException.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 6]
```diff
@@ -1098,6 +1110,20 @@ private boolean enableMergingCustomClaimMappingsWithDefaultMappings() {
...
 
+    /**
+     * Handle role app association in roles claim.
+     *
+     * @param appAssociatedRoles App associated roles.
+     * @param mappedAttrs Mapped claim attributes.
+     */
+    private void handleRoleAppAssoication(Map<String, String> mappedAttrs, String appAssociatedRoles) {
+
+        if (mappedAttrs.containsKey(getLocalGroupsClaimURI())) {
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:1121]: assign getLocalGroupsClaimURI() to a variable and reuse

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Get the groups of the local authenticated user.
+     *
+     * @param authenticatedUser Authenticated user.
+     * @return Groups of the local user.
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java:256]: exception is not in comment

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java (hunk 2 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java]: https://github.com/wso2/carbon-identity-framework/pull/5070/commits/cf43977ae1c82e009dda795205668047624a0925

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java (hunk 3 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java]: https://github.com/wso2/carbon-identity-framework/pull/5070/commits/cf43977ae1c82e009dda795205668047624a0925

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java (hunk 4 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java]: Fixed with Fixed with https://github.com/wso2/carbon-identity-framework/pull/5070/commits/cf43977ae1c82e009dda795205668047624a0925

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java (hunk 5 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/util/RoleResolverUtils.java]: Fixed with Fixed with https://github.com/wso2/carbon-identity-framework/pull/5070/commits/cf43977ae1c82e009dda795205668047624a0925

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java (hunk 6 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java]: Fixed with Fixed with https://github.com/wso2/carbon-identity-framework/pull/5070/commits/cf43977ae1c82e009dda795205668047624a0925

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java (hunk 7 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java:256]: Fixed with Fixed with https://github.com/wso2/carbon-identity-framework/pull/5070/commits/cf43977ae1c82e009dda795205668047624a0925

**Review by thanujalk — APPROVED** [review]

**Pull request merged**